File: src/java/1 - Two Sum.java
package leetcode;

import java.util.HashMap;

public class Solution {

    public int[] twoSum(int[] nums, int target) {
        return v2(nums, target);
    }

    private int[] v1(int[] nums, int target) {
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[i] + nums[j] == target) {
                    return new int[]{i, j};
                }
            }
        }
        throw new IllegalArgumentException("no solution...");
    }

    private int[] v2(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(target - nums[i])) {
                return new int[]{map.get(target - nums[i]), i};
            }
            map.put(nums[i], i);
        }
        throw new IllegalArgumentException("no solution...");
    }

    private static void print(int[] nums, int target) {
        int[] res = new Solution().twoSum(nums, target);
        System.out.println("[" + res[0] + ", " + res[1] + "]");
    }

    public static void main(String[] args) {
        print(new int[]{2, 7, 11, 15}, 9);
        print(new int[]{3, 2, 4}, 6);
        print(new int[]{3, 3}, 6);
    }
}

File: src/scala/1 - Two Sum.scala
package leetcode

object Solution {

  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    v3(nums, target)
  }

  def v1(nums: Array[Int], target: Int): Array[Int] = {
    for {i <- nums.indices
         j <- (i + 1) until nums.length
         if nums(i) + nums(j) == target} {
      return Array(i, j)
    }

    throw new IllegalArgumentException("no solution")
  }

  def v2(nums: Array[Int], target: Int): Array[Int] = {
    val map = collection.mutable.Map[Int, Int]()
    for (i <- nums.indices) {
      if (map.contains(target - nums(i))) {
        return Array(i, map(target - nums(i)))
      }
      map += nums(i) -> i
    }

    throw new IllegalArgumentException("no solution")
  }

  // 相对前两者，更Scala-Style的实现
  def v3(nums: Array[Int], target: Int): Array[Int] = {
    val map = nums.zipWithIndex.toMap
    nums.zipWithIndex
      .find({ case (num, index) => map.contains(target - num) && map(target - num) != index })
      .map({ case (num, index) => Array(index, map(target - num))})
      .get
  }

  def main(args: Array[String]): Unit = {
    @inline def print(nums: Array[Int], target: Int): Unit = {
      println(twoSum(nums, target).mkString("[", ", ", "]"))
    }

    print(Array[Int](2, 7, 11, 15), 9)
    print(Array[Int](3, 2, 4), 6)
    print(Array[Int](3, 3), 6)

  }
}
